pennylane/decomposition/decomposition_graph.py

```diff
@@ -0,0 +1,243 @@
+        for op in all_ops:
+            op_node_idx = self._recursively_add_op_node(op)
+            self._graph.add_edge(op_node_idx, d_node_idx, (op_node_idx, d_node_idx))
+        return d_node_idx
+
+    def solve(self, lazy=True):
+        """Solves the graph using the dijkstra search algorithm.
+
+        Args:
+            lazy (bool): If True, the dijkstra search will stop once optimal decompositions are
+                found for all operations that we originally want to decompose. Otherwise, the
+                entire graph will be explored.
+
+        """
+        self._visitor = _DecompositionSearchVisitor(self._graph, self._original_ops_indices, lazy)
+        start = self._graph.add_node("dummy")
+        self._graph.add_edges_from(
+            [(start, op_node_idx, 1) for op_node_idx in self._target_gate_indices]
+        )
+        rx.dijkstra_search(
+            self._graph,
+            source=[start],
+            weight_fn=self._visitor.edge_weight,
+            visitor=self._visitor,
+        )
+        self._graph.remove_node(start)
+        if self._visitor.unsolved_op_indices:
+            unsolved_ops = [self._graph[op_idx] for op_idx in self._visitor.unsolved_op_indices]
+            op_names = set(op.op_type.__name__ for op in unsolved_ops)
+            raise DecompositionError(
+                f"Decomposition not found for {op_names} to the gate set {self._target_gate_set}"
+            )
+
+    def resource_estimates(self, op) -> Resources:
+        """Returns the resource estimates for a given operator.
+
+        Args:
+            op (Operator): The operator for which to return the resource estimates.
+
+        Returns:
+            Resources: The resource estimates.
+
+        """
+        op_node = resource_rep(type(op), **op.resource_params)
+        op_node_idx = self._op_node_indices[op_node]
+        return self._visitor.d[op_node_idx]
+
+    def decomposition(self, op) -> DecompositionRule:
+        """Returns the optimal decomposition rule for a given operator.
+
+        Args:
+            op (Operator): The operator for which to return the optimal decomposition.
+
+        Returns:
+            DecompositionRule: The optimal decomposition.
+
+        """
+        op_node = resource_rep(type(op), **op.resource_params)
+        op_node_idx = self._op_node_indices[op_node]
+        d_node_idx = self._visitor.p[op_node_idx]
+        return self._graph[d_node_idx].rule
+
+
+class _DecompositionSearchVisitor(DijkstraVisitor):
+    """The visitor used in the dijkstra search for the optimal decomposition."""
+
+    def __init__(self, graph: rx.PyDiGraph, original_op_indices: set[int], lazy: bool = True):
+        self._graph = graph
+        self._lazy = lazy
+        self.d: dict[int, Resources] = {}  # maps node indices to the optimal resource estimates
+        self.p: dict[int, int] = {}  # maps operator nodes to the optimal decomposition nodes
```

Comment:
What do the letters `d` and `p` represent? I suppose we can't change those? :grimacing:

Reply:
Short for "distance" and "predecessor". I can totally change them.

Reply:
Done.

Reply:
Great, thanks!